### PR TITLE
add CDN piece download cost

### PR DESCRIPTION
**Pull request by 244372610**
Signed-off-by: sunwp <244372610@qq.com>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist:

- [ ] Code compiles correctly.
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by gaius-qi — CHANGES_REQUESTED** [review]
Refer to the `PieceResult` field.

**Review by jim3ma — APPROVED** [review]
LGTM

File: cdn/rpcserver/rpcserver.go [hunk 1]
```diff
@@ -95,12 +95,13 @@ func (css *Server) ObtainSeeds(ctx context.Context, req *cdnsystem.SeedRequest,
...
-				PieceMd5:    piece.PieceMd5,
-				PieceOffset: piece.OriginRange.StartIndex,
-				PieceStyle:  piece.PieceStyle,
+				PieceNum:     int32(piece.PieceNum),
+				RangeStart:   piece.PieceRange.StartIndex,
+				RangeSize:    piece.PieceLen,
+				PieceMd5:     piece.PieceMd5,
+				PieceOffset:  piece.OriginRange.StartIndex,
+				PieceStyle:   piece.PieceStyle,
+				DownloadCost: piece.DownloadCost,
```
Comment (gaius-qi) [inline cdn/rpcserver/rpcserver.go:104]: `PieceInfo` needs to be consistent with the `PieceResult` of  `scheduler.proto`.

https://github.com/dragonflyoss/Dragonfly2/blob/main/pkg/rpc/scheduler/scheduler.proto#L96
https://github.com/dragonflyoss/Dragonfly2/blob/main/pkg/rpc/scheduler/scheduler.proto#L98

File: cdn/rpcserver/rpcserver.go [hunk 2]
```diff
@@ -189,12 +190,13 @@ func (css *Server) GetPieceTasks(ctx context.Context, req *base.PieceTaskRequest
...
```
Comment (gaius-qi) [inline cdn/rpcserver/rpcserver.go:199]: `PieceInfo` needs to be consistent with the `PieceResult` of  `scheduler.proto`.

https://github.com/dragonflyoss/Dragonfly2/blob/main/pkg/rpc/scheduler/scheduler.proto#L96
https://github.com/dragonflyoss/Dragonfly2/blob/main/pkg/rpc/scheduler/scheduler.proto#L98

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: cdn/supervisor/cdn/cache_writer.go [hunk 3]
```diff
@@ -139,12 +144,15 @@ loop:
...
 			jobCh <- &piece{
-				taskID:       seedTask.ID,
-				pieceNum:     uint32(curPieceNum),
-				pieceSize:    uint32(seedTask.PieceSize),
-				pieceContent: bb,
+				taskID:            seedTask.ID,
+				pieceNum:          uint32(curPieceNum),
+				pieceSize:         uint32(seedTask.PieceSize),
+				pieceContent:      bb,
+				downloadCost:      uint64(end.Sub(start).Milliseconds()),
```
Comment (gaius-qi) [inline cdn/supervisor/cdn/cache_writer.go:153]: Use `time.Now().UnixNano()`, refer to https://github.com/dragonflyoss/Dragonfly2/blob/main/client/daemon/peer/piece_manager.go#L186

File: pkg/rpc/base/base.pb.validate.go [hunk 4]
```diff
@@ -34,25 +34,62 @@ var (
 )
 
 // Validate checks the field values on GrpcDfError with the rules defined in
-// the proto definition for this message. If any rules are violated, an error
-// is returned.
+// the proto definition for this message. If any rules are violated, the first
+// error encountered is returned, or nil if there are no violations.
 func (m *GrpcDfError) Validate() error {
+	return m.validate(false)
```
Comment (gaius-qi) [inline pkg/rpc/base/base.pb.validate.go:40]: Check your dependency library version.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: cdn/supervisor/cdn/cache_writer.go [hunk 5]
```diff
@@ -139,12 +144,15 @@ loop:
...
+				beginDownloadTime: uint64(start.UnixMilli()),
```
Comment (gaius-qi) [inline cdn/supervisor/cdn/cache_writer.go:154]: Use time.Now().UnixNano(), refer to https://github.com/dragonflyoss/Dragonfly2/blob/main/client/daemon/peer/piece_manager.go#L186

File: cdn/supervisor/cdn/cache_writer.go [hunk 6]
```diff
@@ -139,12 +144,15 @@ loop:
...
+				endDownloadTime:   uint64(end.UnixMilli()),
```
Comment (gaius-qi) [inline cdn/supervisor/cdn/cache_writer.go:155]: Use time.Now().UnixNano(), refer to https://github.com/dragonflyoss/Dragonfly2/blob/main/client/daemon/peer/piece_manager.go#L186

**Review by gaius-qi — CHANGES_REQUESTED** [review]

**Review by gaius-qi — APPROVED** [review]
LGTM

**Review by gaius-qi — APPROVED** [review]
LGTM

**Review by jim3ma — APPROVED** [review]
LGTM

**Pull request merged**